esi_leap/console/websocketproxy.py

```diff
@@ -0,0 +1,164 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""
+Websocket proxy adapted from similar code in Nova
+"""
+
+from http import HTTPStatus
+import os
+import socket
+import threading
+import traceback
+from urllib import parse as urlparse
+import websockify
+
+from oslo_log import log as logging
+from oslo_utils import importutils
+from oslo_utils import timeutils
+
+from esi_leap.common import ironic
+import esi_leap.conf
+from esi_leap.objects import console_auth_token
+
+
+CONF = esi_leap.conf.CONF
+LOG = logging.getLogger(__name__)
+
+
+# Location of WebSockifyServer class in websockify v0.9.0
+websockifyserver = importutils.try_import("websockify.websockifyserver")
+
+
+class ProxyRequestHandler(websockify.ProxyRequestHandler):
+    def __init__(self, *args, **kwargs):
+        websockify.ProxyRequestHandler.__init__(self, *args, **kwargs)
+
+    def verify_origin_proto(self, connect_info, origin_proto):
+        if "access_url_base" not in connect_info:
+            detail = "No access_url_base in connect_info."
+            raise Exception(detail)
+            # raise exception.ValidationError(detail=detail)
+
+        expected_protos = [urlparse.urlparse(connect_info.access_url_base).scheme]
+        # NOTE: For serial consoles the expected protocol could be ws or
+        # wss which correspond to http and https respectively in terms of
+        # security.
+        if "ws" in expected_protos:
+            expected_protos.append("http")
+        if "wss" in expected_protos:
+            expected_protos.append("https")
+
+        return origin_proto in expected_protos
+
+    def _get_connect_info(self, token):
+        """Validate the token and get the connect info."""
+        connect_info = console_auth_token.ConsoleAuthToken.validate(token)
+        if CONF.serialconsoleproxy.timeout > 0:
+            connect_info.expires = (
+                timeutils.utcnow_ts() + CONF.serialconsoleproxy.timeout
+            )
+
+        # get host and port
+        console_info = ironic.get_ironic_client().node.get_console(
+            connect_info.node_uuid
+        )
+        url = urlparse.urlparse(console_info["console_info"]["url"])
```

Comment:
What does the console url look like in this case, and should we be validating the value somehow (e.g., if the user enabled a console type that we don't support)?

Reply:
I think you have missed the nature of my question; I understand that the console url is controlled by ironic. However, will it differ depending on the enabled console type?

Reply:
Ah - it will if it's web console enabled, in which case the URL will simply be an http URL. So for future-proofing purposes, perhaps we should check if the URL is prefixed with tcp; if not, return an error (since the serial console proxy won't work for a web console).